Sources/NIOPerformanceTester/WebSocketFrameDecoderBenchmark.swift

```diff
@@ -15,36 +15,68 @@
 import NIO
 import NIOWebSocket
 
-
 final class WebSocketFrameDecoderBenchmark {
 
     private let channel: EmbeddedChannel
     private let runCount: Int
+    private let dataSize: Int
+    private var maskKey: WebSocketMaskingKey?
     private var data: ByteBuffer
 
-    init(dataSize: Int, runCount: Int) {
+    init(dataSize: Int, runCount: Int, withMaskKey: Bool = false) {
         self.channel = EmbeddedChannel()
+        self.dataSize = dataSize
         self.data = ByteBufferAllocator().buffer(capacity: dataSize)
         self.runCount = runCount
+        if withMaskKey {
+            self.maskKey = [0x80, 0x08, 0x10, 0x01]
+            self.data.webSocketMask(self.maskKey!)
+        }
     }
 }
 
 extension WebSocketFrameDecoderBenchmark: Benchmark {
 
     func setUp() throws {
-        self.data.writeBytes([0x81, 0x00]) // empty websocket
+        self.data = ByteBufferAllocator().buffer(size: self.dataSize)
         try self.channel.pipeline.addHandler(ByteToMessageHandler(WebSocketFrameDecoder())).wait()
     }
 
     func tearDown() {
         _ = try! self.channel.finish()
     }
 
-    func run() throws  -> Int{
-        for _ in 0..<runCount {
+    func run() throws -> Int {
+        if let maskKey = self.maskKey {
+            return try self.run(with: maskKey)
+        } else {
+            return try self.runWithoutMaskKey()
+        }
+    }
+
+    private func runWithoutMaskKey() throws -> Int {
+        for _ in 0..<self.runCount {
+            try self.channel.writeInbound(self.data)
+            let _: WebSocketFrame? =  try self.channel.readInbound()
+        }
+        return 1
+    }
+
+    private func run(with maskKey: WebSocketMaskingKey) throws -> Int {
+        for _ in 0..<self.runCount {
+            self.data.webSocketUnmask(maskKey)
             try self.channel.writeInbound(self.data)
             let _: WebSocketFrame? =  try self.channel.readInbound()
         }
         return 1
     }
 }
+
+extension ByteBufferAllocator {
+    fileprivate func buffer(size: Int) -> ByteBuffer {
+        var data = self.buffer(capacity: size)
+        let bytes = size / 8
+        data.writeBytes([0x81] + Array(repeatElement(0, count: bytes - 1)))
```

Comment:
So this is not quite right: the protocol doesn't really work this way.

You want something like this:

```swift
fileprivate func buffer(size: Int, maskingKey: WebSocketMaskingKey?) -> ByteBuffer {
    var data = self.buffer(capacity: size)

    // Calculate some information about the mask.
    let maskBitMask: UInt8 = maskingKey != nil ? 0x80 : 0x00

    // Time to add the extra bytes. To avoid checking this twice, we also start writing stuff out here.
    switch size {
    case 0...125:
        data.writeInteger(UInt8(0x81))
        data.writeInteger(UInt8(size) | maskBitMask)
    case 126...Int(UInt16.max):
        data.writeInteger(UInt8(0x81))
        data.writeInteger(UInt8(126) | maskBitMask)
        data.writeInteger(UInt16(size))
    case (Int(UInt16.max) + 1)...:
        data.writeInteger(UInt8(0x81))
        data.writeInteger(UInt8(127) | maskBitMask)
        data.writeInteger(UInt64(size))
    }

    if let maskingKey = maskingKey {
        data.writeBytes(maskingKey)
    }

    data.writeBytes(repeatElement(0, count: size))
    return data
}
```

You then want to have one test for each of the following cases:

- size = 125, masking key = nil
- size = 125, masking key = [0x80, 0x08, 0x10, 0x01]
- size = UInt16.max, maskingKey = nil
- size = UInt16.max, maskingKey = [0x80, 0x08, 0x10, 0x01]
- size = UInt16.max + 1, maskingKey = nil
- size = UInt16.max + 1, maskingKey = [0x80, 0x08, 0x10, 0x01]